### ISSUE TITLE
Dropout applied to pooling layer or convolution layer?

### ISSUE THREAD

**Issue by YunshuLiu** [nagadomi/kaggle-cifar10-torch7#5]
Hi nagadomi,
I am not quite familiar with deep learning and even less for torch, I am wondering in the network described in your kaggle-cifar10-tourch7 repo, dropout is applied to pooling layer or convolution layer?
A more general question is : is it different to put dropout right after convolution layer and put it after maxpooling?

   model:add(ccn2.SpatialConvolution(64, 64, 3, 1, 1))
   model:add(nn.ReLU())
   model:add(ccn2.SpatialMaxPooling(2, 2))
   model:add(nn.Dropout(0.25))
vs. 
   model:add(ccn2.SpatialConvolution(64, 64, 3, 1, 1))
   model:add(nn.ReLU())
   model:add(nn.Dropout(0.25))
   model:add(ccn2.SpatialMaxPooling(2, 2))

Thanks,
Yunshu

**Comment by nagadomi** [comment]
That idea of dropout comes from the Network In Network.
https://github.com/mavenlin/cuda-convnet/blob/master/NIN/cifar-10_def#L50

and, In my opinion, max pooling after dropout is incorrect.
not matter:  \* -> dropout -> linear, \* -> dropoput -> convolution, \* -> dropout -> average pooling
bad: \* -> dropout -> max pooling

**Comment by YunshuLiu** [comment]
Thanks for your answer!

**Comment by Coderx7** [comment]
@nagadomi Can you explain more on this? 
Why do you consider using dropout before any pooling  (or even a convolution) operation a bad thing?

**Comment by nagadomi** [comment]
my opinion is: "dropout->maxpooling" is only bad thing.
But I do not have much confidence.

I've made example code. see this: https://gist.github.com/nagadomi/25fbf503b53697befc45fc3542943e32
result:

``` lua
{
  dropout_avg_conv : 
    {
      train/predict : 0.99867722777529
      train : 64.987333992004
      predict : 65.07341129303
    }
  dropout_conv_max_conv : 
    {
      train/predict : 1.0416783267109
      train : 2198.5091051025
      predict : 2110.5451162109
    }
  dropout_max_conv : 
    {
      train/predict : 1.5319914154708
      train : 156.03075260925
      predict : 101.84832044983
    }
  dropout_conv_conv : 
    {
      train/predict : 0.99925533672022
      train : 8856.2585688477
      predict : 8862.8584140625
    }
}
```

In the case of dropout_max_conv(dropout->maxpooling->convolution), the mean of output values, there is clearly difference between train-time and test-time. During prediction, dropout layer reduces **max value** by the scaling factor. It causes different scaling of the maxpooling's output values between train-time and test-time.

**Comment by Coderx7** [comment]
Thank you very much, are these numbers the elapsed time concerning the training/test execution time?

**Comment by nagadomi** [comment]
Sorry, my english skill is very poor.
Maybe I can't relate what I want to say to you.
So please read the code on the assumption that you understanding dropout,max pooling, average pooling and convolution operations.

_I meant to:_
these numbers: a mean value of output values of small example networks.
train-time: the training process
test-time: the test/prediction process
